hera_cal/apply_cal.py

```diff
@@ -0,0 +1,99 @@
+import numpy as np
+import hera_cal.io as io
+from pyuvdata import UVCal, UVData
+
+def recalibrate_in_place(data, data_flags, new_gains, cal_flags, old_gains=None, gain_convention = 'divide'):
+    '''Update data and data_flags in place, taking out old calibration solutions, putting in
+    new calibration solutions, and updating flags from those calibration solutions. Previously 
+    flagged data is left unmodified. Missing antennas from either the new gains or (if it's not None),
+    the old gains are automatically flagged in the data.
+    
+    Arguments:
+        data: DataContainer containing baseline-pol complex visibility data. This is modified in place.
+        data_flags: DataContainer containing data flags. This is modified in place.
+        new_gains: Dictionary of complex calibration gains to apply with keys like (1,'x')
+        cal_flags: Dictionary with keys like (1,'x') of per-antenna boolean flags to update data_flags
+            if either antenna in a visibility is flagged.
+        old_gains: Dictionary of complex calibration gains to take out with keys like (1,'x').
+            Default of None implies that the data is raw (i.e. uncalibrated).
+        gain_convention: str, either 'divide' or 'multiply'. 'divide' means V_obs = gi gj* V_true,
+            'multiply' means V_true = gi gj* V_obs.
+    '''
+    for (i,j,pol) in data.keys():
+        if not np.all(data_flags[(i,j,pol)]):
+            # if any antenna is missing from the gains, the data is flagged
+            if new_gains.has_key((i,pol[0])) and new_gains.has_key((j,pol[1])) and (old_gains == None
+                or (old_gains.has_key((i,pol[0])) and old_gains.has_key((j,pol[1])))):
+                    gigj_new = new_gains[(i,pol[0])] * np.conj(new_gains[(j,pol[1])])
+                    if old_gains is not None:
+                        gigj_old = old_gains[(i,pol[0])] * np.conj(old_gains[(j,pol[1])])
+                    else:
+                        gigj_old = np.ones_like(gigj_new)
+                    # update all the data, assuming it wasn't flagged to begin with
+                    if gain_convention == 'divide':
+                        data[(i,j,pol)][~data_flags[(i,j,pol)]] *= (gigj_old / gigj_new)[~data_flags[(i,j,pol)]]
```

Comment:
This line should be "/= " instead of "*="?

Reply:
This does assume that the conventions of both the old and new gains are the same. Do we want to support them not being the same?